### DISCUSSION TITLE
Preferred method to load runtime configurations for tests?

### DISCUSSION THREAD

**Discussion by joshuacookdev** [xunit/xunit#2184] [Question] [answered]
Looking to have some paths stored in a config file (that way can easily be edited outside of altering code, and I'm not sure of the best way to handle loading during test running.

I saw the [following example](https://stackoverflow.com/questions/10720198/app-config-for-xunit) on SO, but didn't know if there was a preferred methodology over something like this.

<details>
<summary>Example using ctor/dispose</summary>

```
    public class MyTest : IDisposable
    {
        public IServiceProvider Services { get; private set; }
        public MyProjectOptions Options { get; private set; }
        public Logger Logger { get; private set; }

        private void Configure()
        {
            // appsettings.workspace.json for custom developer configuration
            var configuration = new ConfigurationBuilder()
                .SetBasePath(Directory.GetCurrentDirectory())
                .AddJsonFile("appsettings.json")
                .AddJsonFile("appsettings.workspace.json", optional: true)
                .Build();

            Logger = new LoggerConfiguration()
                .MinimumLevel.Debug()
                .WriteTo.LiterateConsole()
                .WriteTo.RollingFile("logs/{Date}-log.txt")
                .CreateLogger();

            Options = configuration.GetSection("MyProject").Get<MyProjectOptions>();

            var services = new ServiceCollection();
            services.AddSingleton<ILogger>(s => Logger);
            // other DI logic and initializations ...
            //services.AddTransient(x => ...);

            Services = services.BuildServiceProvider();
        }

        public MyTest()
        {
            Configure();

            // ... initialize data in the test database ...
            var data = Services.GetService<TestDataService>();
            data.Clean();
            data.SeedData();
        }

        public void Dispose()
        {
            // ... clean-up data in the test database ...
            var data = Services.GetService<TestDataService>();
            data.Clean();
        }
    }
```
</details>

**Comment by JDCain** [comment, marked as answer]
That should work. You can also implement [IAsyncLifetime](https://github.com/xunit/xunit/blob/master/src/xunit.core/IAsyncLifetime.cs) if you need to handle any async Tasks.

My use case requires others using my base test to be able to add strongly typed settings read from a json file so I allow for the passing in of a generic settings type. A user can then add their own setting type which is then passed down to the json desterilizer.

```
    public abstract class TestBase : TestBase<CommonSettingsModel>
    {
        protected TestBase(ITestOutputHelper output) : base(output)
        {
        }
    }
    public abstract class TestBase<T> where T : ICommonSettings, new()
    {
        protected TestBase()
        {
            Settings = SettingsUtil.GetSettings<T>();
        }

        protected T Settings { get; }
    }

```

**Reply by joshuacookdev** [reply]
I like that, actually.

My current case is basically me setting filepaths for different environments, so I don't really care for async in this scenario. I'll definitely throw this in my back pocket for later, though!

Would you consider a setup like what I laid out as best practice in this situation?

**Reply by bradwilson** [reply]
There is yet another option, which is that you could implement that logic as a [fixture](https://xunit.net/docs/shared-context) which would allow longer lifetimes/reusability, depending on how expensive the startup is (and whether it can be easily reused or needs to be fresh for every test).

**Reply by bradwilson** [reply]
Going to mark this as answered. Please let me know if you have any further questions.
